File: tests/test_drawing_bounds.py
"""Headless sanitizer checks for real drawing, SCR readers and BMP name ownership."""
import argparse
import hashlib
import json
import os
from pathlib import Path
import shutil
import subprocess
import tempfile


def build(root, output, sanitize=True):
    root, output = Path(root), Path(output)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text("unbuilt probe")
    os.utime(output, (1, 1))
    before = output.stat().st_mtime_ns
    command = ["cc", "-std=gnu11", "-DPLATFORM_LINUX", "-O1", "-g", "-no-pie",
               "-ffunction-sections", "-fdata-sections",
               "-I" + str(root / "platform"), "-I" + str(root / "src/engine"),
               str(root / "tests/test_drawing_bounds.c"), str(root / "src/engine/graphics.c"),
               str(root / "src/engine/utils.c"), str(root / "platform/platform_linux.c"),
               "-Wl,--gc-sections", "-Wl,--wrap=malloc", "-Wl,--wrap=calloc", "-Wl,--wrap=free",
               "-lX11", "-lasound", "-pthread", "-o", str(output)]
    if sanitize:
        command += ["-fsanitize=address,undefined", "-fno-sanitize-recover=all"]
    p = subprocess.run(command, capture_output=True, text=True)
    assert p.returncode == 0, p.stdout + p.stderr
    assert output.stat().st_mtime_ns > before, "probe artifact timestamp did not advance"


def run(exe, case, directory, extra=()):
    directory.mkdir(parents=True, exist_ok=True)
    (directory / "dump/SCR").mkdir(parents=True, exist_ok=True)
    p = subprocess.run([str(exe), case, *map(str, extra)], cwd=directory,
                       capture_output=True, text=True, timeout=20)
    output = p.stdout + p.stderr
    assert f"WITNESS graphics.c drawing {case} BEGIN" in output, output
    return p.returncode, output


def copy_source(source, target):
    for folder in ("src/engine", "platform", "tests"):
        (target / folder).mkdir(parents=True, exist_ok=True)
        for path in (source / folder).glob("*"):
            if path.suffix in (".c", ".h"):
                shutil.copyfile(path, target / folder / path.name)


def replace(path, old, new, count=1):
    text = path.read_text()
    assert text.count(old) == count, (str(path), old, text.count(old))
    path.write_text(text.replace(old, new))


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--source", type=Path, default=Path(__file__).resolve().parents[1])
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--phase", choices=["smoke", "regression", "all"], default="all")
    parser.add_argument("--mutations", action="store_true", help="manual rebuilt negative controls")
    args = parser.parse_args()
    args.source, args.output = args.source.resolve(), args.output.resolve()
    args.output.mkdir(parents=True, exist_ok=True)
    exe = args.output / "drawing-probe"
    build(args.source, exe)
    cases = ["circle-fill", "even-scr"] if args.phase == "smoke" else [
        "circle-outline", "empty-bmp-reload", "empty-bmp-alias", "odd-scr", "odd-dump", "even-dump"]
    if args.phase == "all": cases = ["circle-fill", "even-scr", *cases]
    records = []
    for case in cases:
        code, output = run(exe, case, args.output / case)
        if case.startswith("odd-"):
            assert code == 1 and "SCR 'ODD.SCR': unsupported odd width 3" in output, output
            assert "refused before style/decode allocation" in output, output
            assert not (args.output / case / "dump/SCR/ODD.SCR.xpm").exists()
        else:
            assert code == 0 and f"drawing {case} PASS" in output, output
        if case == "even-dump":
            data = (args.output / case / "dump/SCR/EVEN.SCR.xpm").read_bytes()
            assert b'"4 2 16 1"' in data and b'"1234",\n"5678"}' in data
        records.append({"case": case, "result": "PASS", "output": output})
        print(f"PASS src/engine/{'dump.c' if 'dump' in case else 'graphics.c'}: {case}", flush=True)

    parity = None
    if args.phase != "smoke":
        with tempfile.TemporaryDirectory(dir=args.output) as folder:
            legacy = Path(folder)
            copy_source(args.source, legacy)
            replace(legacy / "src/engine/graphics.c", "2 * (x - y)", "((x - y) << 1)", count=2)
            old_exe = legacy / "legacy-probe"
            build(legacy, old_exe, sanitize=False)
            current_pixels, legacy_pixels = legacy / "current.bin", legacy / "legacy.bin"
            for binary, pixels in ((exe, current_pixels), (old_exe, legacy_pixels)):
                code, output = run(binary, "circle-corpus", legacy, (pixels,))
                assert code == 0 and "circle-corpus PASS" in output, output
            current_data = current_pixels.read_bytes()
            assert current_data == legacy_pixels.read_bytes(), "src/engine/graphics.c: circle pixels changed"
            parity = {"byte_count": len(current_data), "sha256": hashlib.sha256(current_data).hexdigest(),
                      "cases": 12, "scales": [1, 2], "byte_identical": True}
            print("PASS src/engine/graphics.c: 12 circle pixel cases equal pre-fix arithmetic", flush=True)

    mutations = []
    if args.mutations:
        definitions = [
            ("circle-fill", "graphics.c",
             "void grDrawCircle(PlatformSurface *sfc, int x1, int y1, int width, int height, uint8 fgColor, uint8 bgColor)\n{",
             "void grDrawCircle(PlatformSurface *sfc, int x1, int y1, int width, int height, uint8 fgColor, uint8 bgColor)\n{\n    return;",
             "circle center pixel"),
            ("circle-fill", "graphics.c", "\n            d += 2 * (x - y) + 5;",
             "\n            d += ((x - y) << 1) + 5;", "left shift of negative value"),
            ("circle-outline", "graphics.c", "\n                d += 2 * (x - y) + 5;",
             "\n                d += ((x - y) << 1) + 5;", "left shift of negative value"),
            ("odd-scr", "graphics.c", "if (scrResource->width % 2)",
             "if (0 && scrResource->width % 2)", "heap-buffer-overflow"),
            ("odd-dump", "dump.c", "if (scrResource->width % 2)",
             "if (0 && scrResource->width % 2)", "odd SCR was accepted"),
            ("empty-bmp-reload", "graphics.c", "        grReleaseBmp(ttmSlot, slotNo);",
             "        if (ttmSlot->numSprites[slotNo]) grReleaseBmp(ttmSlot, slotNo);", "liveCount == 1"),
            ("empty-bmp-alias", "graphics.c", "findBmpResource(ttmSlot->bmpNames[slotNo])",
             "findBmpResource(strArg)", "heap-use-after-free"),
        ]
        for case, filename, old, new, expected in definitions:
            with tempfile.TemporaryDirectory(dir=args.output) as folder:
                mutant = Path(folder)
                copy_source(args.source, mutant)
                replace(mutant / "src/engine" / filename, old, new)
                mutant_exe = mutant / "mutant-probe"
                build(mutant, mutant_exe)
                code, output = run(mutant_exe, case, mutant)
                assert code != 0 and expected in output, f"SURVIVED src/engine/{filename}: {case}\n{output}"
                assert f"drawing {case} PASS" not in output
                mutations.append({"file": f"src/engine/{filename}", "case": case, "result": "FIRED",
                                  "rebuilt": True, "runtime_witness": True, "exit_code": code,
                                  "expected_failure": expected, "output": output})
                print(f"FIRED src/engine/{filename}: {case} (rebuilt and executed)", flush=True)
    report = {"phase": args.phase, "cases": records, "circle_parity": parity, "mutations": mutations}
    (args.output / "drawing-report.json").write_text(json.dumps(report, indent=2))


if __name__ == "__main__":
    main()

File: tests/test_drawing_bounds.c
/* Real graphics.c drawing and ownership checks without a display or archive. */
#undef NDEBUG
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "mytypes.h"
#include "resource.h"
#include "graphics.h"
#include "art_style.h"
#include "../src/engine/dump.c"

static void *allocations[64];
static unsigned liveCount, allocCount, screenLoads, expectedAllocCount;
static int checkingOdd;
void *__real_malloc(size_t);
void *__real_calloc(size_t, size_t);
void __real_free(void *);
static void *track(void *p) {
    assert(p && liveCount < 64);
    allocations[liveCount++] = p;
    allocCount++;
    return p;
}
void *__wrap_malloc(size_t size) { return track(__real_malloc(size)); }
void *__wrap_calloc(size_t count, size_t size) { return track(__real_calloc(count, size)); }
void __wrap_free(void *p) {
    if (!p) return;
    unsigned i;
    for (i = 0; i < liveCount && allocations[i] != p; i++) {}
    assert(i < liveCount);
    allocations[i] = allocations[--liveCount];
    __real_free(p);
}

static struct TScrResource screen;
static struct TBmpResource emptyBmp = {.resName="EMPTY.BMP", .numImages=0};
static struct TPalResource palette;
struct TScrResource *findScrResource(const char *name) {
    assert(!strcmp(name, screen.resName)); return &screen;
}
struct TBmpResource *findBmpResource(const char *name) {
    assert(!strcmp(name, emptyBmp.resName)); return &emptyBmp;
}
PlatformSurface *artStyleLoadScreen(const char *name, int w, int h) {
    (void)name; (void)w; (void)h; screenLoads++; return NULL;
}
PlatformSurface *artStyleLoadSprite(const char *name, int i, int w, int h) {
    (void)name; (void)i; (void)w; (void)h; assert(!"empty BMP must not load a sprite"); return NULL;
}
void artStyleReportUsage(void) {}
void islandRelease(void) {}

static void oddCleanup(void) {
    if (checkingOdd) {
        assert(allocCount == expectedAllocCount && screenLoads == 0 && !grBackgroundSfc);
        puts("WITNESS graphics.c odd SCR refused before style/decode allocation");
        free(screen.uncompressedData);
        assert(liveCount == 0);
    }
}
static void evenScreen(void) {
    static uint8 data[] = {0x12, 0x34, 0x56, 0x78};
    screen = (struct TScrResource){.resName="EVEN.SCR", .width=4, .height=2,
        .uncompressedSize=sizeof(data), .uncompressedData=data};
    for (int scale = 1; scale <= 2; scale++) {
        grScale = scale;
        grLoadScreen(screen.resName);
        assert(platformGetSurfaceWidth(grBackgroundSfc) == 4 * scale);
        assert(platformGetSurfaceHeight(grBackgroundSfc) == 2 * scale);
        uint8 *pixels = platformGetSurfacePixels(grBackgroundSfc);
        int pitch = platformGetSurfacePitch(grBackgroundSfc);
        for (int y = 0; y < 2 * scale; y++) {
            for (int x = 0; x < 4 * scale; x++) {
                int index = (y / scale) * 4 + x / scale + 1;
                uint8 *p = pixels + y * pitch + x * 4;
                assert(p[0] == palette.colors[index].b * 4);
                assert(p[1] == palette.colors[index].g * 4);
                assert(p[2] == palette.colors[index].r * 4 && p[3] == 255);
            }
        }
        graphicsEnd();
        assert(liveCount == 0);
    }
}
static void circle(int outline) {
    PlatformSurface *surface = platformCreateSurface(640, 480);
    grDrawCircle(surface, 100, 80, 40, 40, outline ? 2 : 1, 1);
    uint8 *pixels = platformGetSurfacePixels(surface);
    int pitch = platformGetSurfacePitch(surface);
    uint8 *center = pixels + 99 * pitch + 119 * 4;
    uint8 *edge = pixels + 80 * pitch + 119 * 4;
    uint8 *outside = pixels + 79 * pitch + 99 * 4;
    assert(center[0] == 4 && center[1] == 8 && center[2] == 12 && center[3] == 255
           && "circle center pixel");
    assert(edge[0] == (outline ? 8 : 4) && edge[1] == (outline ? 16 : 8)
           && edge[2] == (outline ? 24 : 12) && edge[3] == 255 && "circle edge pixel");
    assert(outside[0] == 0 && outside[1] == 0 && outside[2] == 0 && outside[3] == 0
           && "circle outside pixel");
    grFreeLayer(surface);
}
static void circleCorpus(const char *path) {
    /* Filled/outlined, small/large, clipping/offset and both shipping scales. */
    static const int shapes[][6] = {
        {100, 80, 40, 1, 1, 0}, {100, 80, 40, 2, 1, 0},
        {-80, -160, 800, 3, 4, 0}, {630, 470, 40, 5, 6, 0},
        {150, 120, 200, 7, 8, -272}, {638, 478, 2, 9, 9, 0}
    };
    FILE *out = fopen(path, "wb");
    assert(out);
    for (int scale = 1; scale <= 2; scale++) {
        grScale = scale;
        for (unsigned i = 0; i < sizeof(shapes) / sizeof(shapes[0]); i++) {
            const int *s = shapes[i];
            PlatformSurface *surface = platformCreateSurface(640 * scale, 480 * scale);
            grDx = s[5]; grDy = 0;
            grDrawCircle(surface, s[0], s[1], s[2], s[2], (uint8)s[3], (uint8)s[4]);
            size_t bytes = (size_t)platformGetSurfacePitch(surface) * 480 * scale;
            assert(fwrite(platformGetSurfacePixels(surface), 1, bytes, out) == bytes);
            grFreeLayer(surface);
        }
    }
    assert(fclose(out) == 0);
}
int main(int argc, char **argv) {
    assert(argc >= 2);
    printf("WITNESS graphics.c drawing %s BEGIN\n", argv[1]); fflush(stdout);
    for (int i = 0; i < 16; i++) {
        palette.colors[i] = (struct TColor){(uint8)(i * 3), (uint8)(i * 2), (uint8)i};
    }
    grLoadPalette(&palette);
    grScale = 1;
    if (!strcmp(argv[1], "circle-fill")) circle(0);
    else if (!strcmp(argv[1], "circle-outline")) circle(1);
    else if (!strcmp(argv[1], "circle-corpus")) {
        assert(argc == 3); circleCorpus(argv[2]);
    } else if (!strcmp(argv[1], "even-scr")) evenScreen();
    else if (!strcmp(argv[1], "odd-scr") || !strcmp(argv[1], "odd-dump")) {
        screen = (struct TScrResource){.resName="ODD.SCR", .width=3, .height=1,
            .uncompressedSize=1, .uncompressedData=malloc(1)};
        screen.uncompressedData[0] = 0x12;
        grHdEnabled = 1;
        checkingOdd = 1;
        expectedAllocCount = allocCount;
        atexit(oddCleanup);
        if (!strcmp(argv[1], "odd-scr")) grLoadScreen(screen.resName);
        else dumpScr(&screen, &palette);
        assert(!"odd SCR was accepted");
    } else if (!strcmp(argv[1], "even-dump")) {
        static uint8 data[] = {0x12, 0x34, 0x56, 0x78};
        screen = (struct TScrResource){.resName="EVEN.SCR", .width=4, .height=2,
            .uncompressedSize=sizeof(data), .uncompressedData=data};
        dumpScr(&screen, &palette);
    } else if (!strcmp(argv[1], "empty-bmp-reload") || !strcmp(argv[1], "empty-bmp-alias")) {
        struct TTtmSlot slot = {0};
        grLoadBmp(&slot, 0, "EMPTY.BMP");
        assert(liveCount == 1 && allocCount == 1 && slot.numSprites[0] == 0);
        const char *requested = !strcmp(argv[1], "empty-bmp-alias") ? slot.bmpNames[0] : "EMPTY.BMP";
        grLoadBmp(&slot, 0, requested);
        assert(liveCount == 1 && allocCount == 2 && !strcmp(slot.bmpNames[0], "EMPTY.BMP"));
        puts("WITNESS graphics.c consecutive zero-image load retains one name");
        grReleaseBmp(&slot, 0);
    } else assert(!"unknown case");
    assert(liveCount == 0);
    printf("WITNESS graphics.c drawing %s PASS\n", argv[1]);
    return 0;
}
